Integrate multi_tensor zero_grad into Optimizer base class

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* #69981
* #69980
* __->__ #69936

Currently, the optimizers in `torch/optim/_multi_tensor/` all override the base Optimizer class' implementation of `zero_grad` with the same foreach zero_grad implementation (e.g. [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/adadelta.py#L93-L114)). There is a TODO that indicates that this should be refactored to the base class once the foreach ops are in good shape. This PR is intended to address that TODO.

Differential Revision: [D33346748](https://our.internmc.facebook.com/intern/diff/D33346748)

mikaylagawarecki committed:
Integrate multi_tensor zero_grad into Optimizer base class

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate multi_tensor zero_grad into Optimizer base class"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate multi_tensor zero_grad into Optimizer base class"


Currently, the optimizers in `torch/optim/_multi_tensor/` all override the base Optimizer class' implementation of `zero_grad` with the same foreach zero_grad implementation (e.g. [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/adadelta.py#L93-L114)). There is a TODO that indicates that this should be refactored to the base class once the foreach ops are in good shape. This PR is intended to address that TODO.

[ghstack-poisoned]

albanD commented on torch/optim/optimizer.py:
```diff
@@ -201,8 +201,12 @@ def zero_grad(self, set_to_none: bool = False):
                 (in one case it does the step with a gradient of 0 and in the other it skips
                 the step altogether).
         """
+        foreach = self.defaults['foreach'] if ('foreach' in self.defaults) else False
```
`foreach = self.defaults.get('foreach', False)`

albanD commented on torch/optim/optimizer.py:
```diff
@@ -214,7 +218,14 @@ def zero_grad(self, set_to_none: bool = False):
+                                p.grad.zero_()
+                            else:
+                                per_device_and_dtype_grads[p.grad.device][p.grad.dtype].append(p.grad)
+            if (foreach):
```
nit no parenthesis

albanD reviewed (commented):
change looks good. just some nits for the impl

mikaylagawarecki committed:
Update on "Integrate multi_tensor zero_grad into Optimizer base class"


Currently, the optimizers in `torch/optim/_multi_tensor/` all override the base Optimizer class' implementation of `zero_grad` with the same foreach zero_grad implementation (e.g. [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/adadelta.py#L93-L114)). There is a TODO that indicates that this should be refactored to the base class once the foreach ops are in good shape. This PR is intended to address that TODO.

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

mikaylagawarecki committed:
Update on "Integrate multi_tensor zero_grad into Optimizer base class"


Currently, the optimizers in `torch/optim/_multi_tensor/` all override the base Optimizer class' implementation of `zero_grad` with the same foreach zero_grad implementation (e.g. [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/adadelta.py#L93-L114)). There is a TODO that indicates that this should be refactored to the base class once the foreach ops are in good shape. This PR is intended to address that TODO.

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate multi_tensor zero_grad into Optimizer base class"


Currently, the optimizers in `torch/optim/_multi_tensor/` all override the base Optimizer class' implementation of `zero_grad` with the same foreach zero_grad implementation (e.g. [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/adadelta.py#L93-L114)). There is a TODO that indicates that this should be refactored to the base class once the foreach ops are in good shape. This PR is intended to address that TODO.

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33346748).

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33346748).
